### PR TITLE
develop

### PR DESCRIPTION
**Pull request by birdierabi**

### CODE REVIEW

File: src/components/game/game.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    await card.flipToFront();
+
+    if (!this.activeCard) {
+      this.activeCard = card;
+      this.isAnimation = false;
+      return;
+    }
+
+    if (this.activeCard.image !== card.image) {
+      this.activeCard.showError();
```
Comment (za3k) [inline src/components/game/game.ts:53]: Add a small comment for each of these three cases. Ex "First flip, second flip doesn't match first; second flip matches first" along with what happens in each case in plain english. For example, "Second flip doesn't match first--wait a few seconds for the user to look at the cards, then flip both face down".

Somewhere in your documentation, explain how Memory works.

File: public/images.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
+[
+  {
+    "category": "unsorted",
```
Comment (za3k) [inline public/images.json:3]: why do you have categories?

File: src/components/game/game.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    super('main');
+    this.cardsField = new CardsField();
+    this.element.appendChild(this.cardsField.element);
+  }
+
+  newGame(images: string[]): void {
+    this.cardsField.clear();
+    const cards = images
+      .concat(images)
+      .map((url) => new Card(url))
```
Comment (za3k) [inline src/components/game/game.ts:25]: It looks like you are using `=>` instead of `->` everywhere. Do you understand the difference yet? If you don't yet that's OK, and this is probably fine.

**Review by za3k — APPROVED** [review]
Overall this looks great! You could use maybe 10 lines of comments and a link to play the game, otherwise I think it looks good as-is. I've added some more suggestions, but that's all they are--suggestions on how you could improve things if you wanted to, not anything that actually needs done.

This is very carefully written code, nice job. I wasn't sure how to actually check out and play the game though, sorry, so I can't comment on the CSS as much.

I was a little taken aback at the architecture--I had to read 32 files in order to review what I felt was about 10-20 lines of actual logic. This isn't necessarily anything you did wrong, I'm just not used to (and don't particularly like) this object-oriented, very split out style. You may want a more modern reviewer, I'm not sure, but I'm happy to help as long as you're happy with my reviews.

The large number of files and spread-out logic does mean you should write better central documentation. Your documentation for programmers (yes I know, that's just you, but just for practice) doesn't need to be in the main github readme, although it can be. It does need to be findable, so think about how you might look for documentation as a developer.
- What files do what? If I'm looking for the XXX logic, what file should I read?
- What does each class do? Add a 1-line comment explaining what each class does.
- Add a 1-line comment explaining what any methods or functions do. API docs or python docstrings are often good examples of how to document a function. The important part is to describe what it does to a USER (what to pass it, and what it returns), not how it works. For example a `shuffle` function should be documented as "returns a copy of the passed-in list shuffled in random order", not "sorts the list using random values"
- Add a 1-line comment explaining what any data members do which are not completely self-explanatory.

Your documentation for users should be prominent and visible. It should include
- What is Memory, and how do you play it?
- How do you try out your Memory game (a link)
- Since this is a work in progress, what works and doesn't (you did a pretty good job of this already, just make sure to keep it up to date)

File: src/components/base-component.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+export class BaseComponent {
+  readonly element: HTMLElement;
+
+  constructor(tag: keyof HTMLElementTagNameMap = 'div', styles: string[] = []) {
+    this.element = document.createElement(tag);
+    if (styles.length) this.element.classList.add(...styles);
```
[the remark this answers is not in the document]
Comment (birdierabi) [reply src/components/base-component.ts:6]: I added `styles.length` check here because `this.element.classList.add(...styles);` will add unnecessary "class" word in DOM. It would be more correct to do something like that:
```ts
constructor(tag: keyof HTMLElementTagNameMap = 'div', styles: string[] | null) {
   this.element = document.createElement(tag);
   if (styles) this.element.classList.add(...styles);
}
```

File: src/components/card/card.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
+import './card.scss';
+import { BaseComponent } from '../base-component';
+
+const FLIP_CLASS = 'card-container--flipped';
+const ERROR_CLASS = 'card-container--error';
+const MATCH_CLASS = 'card-container--success';
+
+export class Card extends BaseComponent {
+  isFlipped = false;
```
[the remark this answers is not in the document]
Comment (birdierabi) [reply src/components/card/card.ts:9]: Yes, it’s possible in TypeScript syntax. But I also agree that I need to fix it.
```ts
export class Card extends BaseComponent {
  isFlipped: boolean;

  constructor(readonly image: string) {
    super('div', ['card-container']);

    this.isFlipped = false;
    
    ...
  }
```

File: src/components/card/card.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+  showError(): void {
+    this.element.classList.add(ERROR_CLASS);
+  }
+
+  deleteError(): void {
+    this.element.classList.remove(ERROR_CLASS);
+  }
+
+  showMatch(): void {
```
[the remark this answers is not in the document]
Comment (birdierabi) [reply src/components/card/card.ts:30]: OK

File: src/components/card/card.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+  flipToFront(): Promise<void> {
+    this.isFlipped = false;
+    return this.flip();
+  }
+
+  private flip(isFront = false): Promise<void> {
+    return new Promise((resolve) => {
+      this.element.classList.toggle(FLIP_CLASS, isFront);
+      this.element.addEventListener('transitionend', () => resolve(), {
```
[the remark this answers is not in the document]
Comment (birdierabi) [reply src/components/card/card.ts:47]: I can swap these two lines. The second argument here is just a handler – a function that runs in case of an event (transitionend – when a CSS animation finishes).
```ts
private flip(isFront = false): Promise<void> {
    return new Promise((resolve) => {
      this.element.addEventListener('transitionend', () => resolve(), {
        once: true,
      });
      this.element.classList.toggle(FLIP_CLASS, isFront);
    });
}
```

File: src/components/cards-field/cards-field.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
+import './cards-field.scss';
+import { BaseComponent } from '../base-component';
+import { Card } from '../card/card';
+
+const SHOW_TIME = 5;
```
[the remark this answers is not in the document]
Comment (birdierabi) [reply src/components/cards-field/cards-field.ts:5]: OK

File: src/components/game/game.ts [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+      .sort(() => Math.random() - 0.5);
```
[the remark this answers is not in the document]
Comment (birdierabi) [reply src/components/game/game.ts:26]: OK

File: src/components/game/game.ts [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+import { BaseComponent } from '../base-component';
+import { Card } from '../card/card';
+import { CardsField } from '../cards-field/cards-field';
+
+const FLIP_DELAY = 1000;
+
+export class Game extends BaseComponent {
+  private readonly cardsField: CardsField;
+
+  private activeCard?: Card;
```
[the remark this answers is not in the document]
Comment (birdierabi) [reply src/components/game/game.ts:11]: OK

File: src/components/game/game.ts [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+  constructor() {
+    super('main');
+    this.cardsField = new CardsField();
+    this.element.appendChild(this.cardsField.element);
+  }
+
+  newGame(images: string[]): void {
+    this.cardsField.clear();
+    const cards = images
+      .concat(images)
```
[the remark this answers is not in the document]
Comment (birdierabi) [reply src/components/game/game.ts:24]: OK

File: src/components/game/game.ts [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+      .sort(() => Math.random() - 0.5);
+
+    cards.forEach((card) => {
+      card.element.addEventListener('click', () => this.cardHandler(card));
+    });
+
+    this.cardsField.addCards(cards);
+  }
+
+  private async cardHandler(card: Card) {
```
[the remark this answers is not in the document]
Comment (birdierabi) [reply src/components/game/game.ts:35]: Thank you, I will read about it.

File: src/components/game/game.ts (hunk 1 above)
Comment (birdierabi) [reply src/components/game/game.ts:53]: OK

File: public/images.json (hunk 2 above)
Comment (birdierabi) [reply public/images.json:3]: Because I will add other types of cards. On the settings page, the user can change the type of cards that will be in the `category` property (for example, animals, sports). Here I need to rename the value of `category` to `sports`.

File: src/components/game/game.ts (hunk 3 above)
Comment (birdierabi) [reply src/components/game/game.ts:25]: here `=>` is an arrow function (the right side is an expression: the function evaluates it and returns the result) [About arrow functions](https://javascript.info/arrow-functions-basics)

File: src/components/card/card.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (za3k) [reply src/components/card/card.ts:47]: Change to: `this.element.addEventListener('transitionend', resolve, { once: true };`.
`() => someFunction()` is redundant.

File: src/components/game/game.ts (hunk 3 above)
Comment (za3k) [reply src/components/game/game.ts:25]: Yes, I know what arrow functions are. :)

**Comment by za3k** [comment]
How good is your English? Did you you have an easy or difficult time with this review? I can try to make things easier to understand if that would help with review.

It is good to improve your English as a programmer, especially if you want a job working for english-speakers.

**Comment by birdierabi** [comment]
> How good is your English? Did you you have an easy or difficult time with this review? I can try to make things easier to understand if that would help with review.
> 
> It is good to improve your English as a programmer, especially if you want a job working for english-speakers. But maybe a review is not the place to learn :)

I have an easy time on it. I admit that I am prone to mistakes. Also, I think I don't have strong writing skills.

I agree with you, code review is definitely not a place for my English improvement. Thank you a lot for your contribution :) I am planning to improve my English for sure.
